client/grafana-data-source/src/lib.rs

```diff
@@ -24,35 +24,51 @@
 //! [`grafana-json-data-source`]: https://github.com/simPod/grafana-json-datasource
 
 use lazy_static::lazy_static;
-use std::collections::HashMap;
 use parking_lot::RwLock;
 
 mod types;
 mod server;
-mod util;
 #[cfg(not(target_os = "unknown"))]
 mod networking;
+mod database;
 
+use database::Database;
 pub use server::run_server;
-pub use util::now_millis;
-
-type Metrics = HashMap<String, Vec<(f32, i64)>>;
+use std::num::TryFromIntError;
 
 lazy_static! {
-	/// The `RwLock` wrapping the metrics. Not intended to be used directly.
-	#[doc(hidden)]
-	pub static ref METRICS: RwLock<Metrics> = RwLock::new(Metrics::new());
+	// The `RwLock` wrapping the metrics database.
+	static ref DATABASE: RwLock<Database> = RwLock::new(Database::new());
 }
 
 /// Write metrics to `METRICS`.
 #[macro_export]
 macro_rules! record_metrics(
 	($($key:expr => $value:expr),*) => {
-		use $crate::{METRICS, now_millis};
-		let mut metrics = METRICS.write();
-		let now = now_millis();
-		$(
-			metrics.entry($key).or_insert_with(Vec::new).push(($value as f32, now));
-		)*
+		$crate::record_metrics_slice(&[
+			$( ($key, $value as f32), )*
+		]);
 	}
 );
+
+/// Write metrics to `METRICS` as a slice.
+pub fn record_metrics_slice(metrics: &[(&'static str, f32)]) -> Result<(), Error> {
```

Comment:
this could be more generic by taking: `IntoIterator<Item = (&'static str, f32)>` instead

Reply:
I might just take the `pub` off actually.

Reply:
The type signature would be `IntoIterator<Item = &'a (&'static str, f32)>` to make it work with slices, which is less nice. I chose to clarify when it should be used instead.

Reply:
fair enough